#391 fixed race condition around open and close

korydraughn opened the pull request:
@michael-conway Please take a moment to look over this code before merging it.

My original fix I discussed with you earlier exposed another issue that this PR resolves.

After moving the handling of cache entry removal out of `closeReplicaToken()`, I discovered that the call to `claimExistingReplicaToken()` was throwing an exception because the cache entry didn't exist anymore. This can happen when `obtainReplicaTokenLock()` is called by one thread and another thread, operating on the same replica, removes the cache entry owning the lock returned by `obtainReplicaTokenLock()`.

This PR addresses this situation by introducing an atomic reference counter. The counter tracks how many threads have obtained, but not acquired, a lock for a particular cache entry. With that, we only allow removal of the cache entry if the atomic reference counter reaches zero.

trel commented on jargon-core/src/main/java/org/irods/jargon/core/connection/ReplicaTokenCacheManager.java:
```diff
@@ -195,12 +210,59 @@ public boolean closeReplicaToken(final String logicalPath, final String userName
+	 * invoked {@code obtainReplicaTokenLock} on the same cache entry).
+	 *
+	 * @param logicalPath {@code String} with the logical path to the file to be
+	 *                    closed * @param userName {@code String} which is the user
```
need to break the userName param to next line

korydraughn replied:
done.

trel commented on jargon-core/src/main/java/org/irods/jargon/core/connection/ReplicaTokenCacheManager.java:
```diff
@@ -195,12 +210,59 @@ public boolean closeReplicaToken(final String logicalPath, final String userName
+		ReplicaTokenCacheKey cacheKey = ReplicaTokenCacheKey.instance(logicalPath, userName);
+
+		ReplicaTokenCacheEntry replicaTokenCacheEntry = replicaTokenCache.get(cacheKey);
+		if (replicaTokenCacheEntry.getOpenCount() <= 0) {
```
why <= ?    can it go negative?   if so, does that mean something else bad happened?

korydraughn replied:
That shouldn't be possible. I am just paranoid.

Perhaps checking for zero is better as that can expose unknown bugs in the logic.

korydraughn replied:
updated to check for zero.

trel replied:
should you have an else() that catches the <= case?   and gets very angry?

korydraughn replied:
I'll add a log statement for that case. Whatever happens after that is on the developer to solve because the log reported that they are holding things incorrectly.

korydraughn replied:
added log statement.

trel replied:
sold.

korydraughn commented:
Disabling the removal of the cache entry didn't fix the issue. However, synchronizing the calls to open/close in the write operation of NFSRODS did resolve the issue.

I'll continue to review this implementation and post updates here.

trel commented:
so it's fixable today, but depends on the client application, rather than jargon to do so.    which... could be an update later if we wanted to get this out.

korydraughn commented:
Correct

korydraughn commented:
I think we have a working solution now.

Important Changes:
- Changed interface and implementation of `ReplicaTokenCacheManager`
  - This caused a slight change in how it is invoked in open and close
- Introduced a new inner class called `ReplicaTokenLock` (can probably be replaced with a normal lock)
- Pulled the lock out of the cache entry
  - Locks now live in a separate hash map inside the cache manager
  - Replica token cache entries can be safely removed based on the open count
  - While externalizing the locks is necessary, this creates a situation around how to clean up unused locks
    - Do we use a background thread for this?
    - Do we clean up the locks on certain function calls?
    - etc.

With the PR presented, the BATS test suite succeeds and parallel writes do not result in `HIERARCHY_ERROR`s. I've run the `dd`-based BATS test, `copy large file`, 100+ times using a for-loop and the `HIERARCHY_ERROR` did not show up. The error normally appears after 2 to 5 runs.

All that is left is a review and whitespace formatting (i.e. indentation).

trel commented on jargon-core/src/main/java/org/irods/jargon/core/pub/io/IRODSFile.java:
```diff
@@ -359,11 +359,11 @@ void close(boolean updateSize, boolean updateStatus, boolean computeChecksum, bo
 
 	/**
-	 * Set the resource token if one is available (dependant on iRODS version)
+	 * Set the replica token if one is available (dependant on iRODS version)
```
dependant  -> dependent

trel commented on jargon-core/src/main/java/org/irods/jargon/core/pub/io/IRODSFileImpl.java:
```diff
@@ -1567,26 +1559,26 @@ public void setIrodsFileSystemAO(final IRODSFileSystemAO irodsFileSystemAO) {
 
 	/**
-	 * Set the resource token if one is available (dependant on iRODS version)
+	 * Set the replica token if one is available (dependant on iRODS version)
```
dependant -> dependent

trel reviewed (commented):
and whitespace, like you said.
